File: app/src/main/java/com/asteph11/finalproject/models/data/TurnHandler.java
package com.asteph11.finalproject.models.data;

public class TurnHandler {

    private int turnCount = 0;

    private int current = 1;
    private Player[] players;

    public TurnHandler(Player p1, Player p2) {
        this.players = new Player[]{p1, p2};
    }

    private int swapIndex() {
        current = ((current+1) % 2);

        if(current == 1) {
            turnCount++;
        }

        return current;
    }

    public Player swapTurns() {
        swapIndex();

        return players[current];
    }

    public int getTurnCount() {
        return turnCount;
    }

    public Player getCurrentPlayer() {
        return players[current];
    }

    public Player getOtherPlayer() {
        return players[((current+1) % 2)];
    }

    public Player getP1() {
        return players[0];
    }

    public Player getP2() {
        return players[1];
    }

    public String toString() {
        return "Current Player: [" + current + "] " + players[current];
    }

    public boolean isReady() {
        if(players == null || players[0] == null || players[1] == null) {
            return false;
        }

        return getP1().isReady() && getP2().isReady();
    }

    public boolean isFinished() {
        isReady();
        return (getP1().isDefeated() || getP2().isDefeated());
    }
}

File: app/src/main/java/com/asteph11/finalproject/models/data/TurnHandler.kt
package com.asteph11.finalproject.models.data

class TurnHandler(p1: Player?, p2: Player?) {
    var turnCount = 0
        private set
    private var current = 1
    private val players: Array<Player?>?
    private fun swapIndex(): Int {
        current = (current + 1) % 2
        if (current == 1) {
            turnCount++
        }
        return current
    }

    fun swapTurns(): Player? {
        swapIndex()
        return players!![current]
    }

    val currentPlayer: Player?
        get() = players!![current]
    val otherPlayer: Player?
        get() = players!![(current + 1) % 2]
    val p1: Player?
        get() = players!![0]
    val p2: Player?
        get() = players!![1]

    override fun toString(): String {
        return "Current Player: [" + current + "] " + players!![current]
    }

    val isReady: Boolean
        get() = if (players == null || players[0] == null || players[1] == null) {
            false
        } else p1!!.isReady && p2!!.isReady
    val isFinished: Boolean
        get() {
            isReady
            return p1!!.isDefeated || p2!!.isDefeated
        }

    init {
        players = arrayOf(p1, p2)
    }
}